Add support for signed values?

```go

func TestBitfieldSimple(t *testing.T) {

	value := int32(1)
	type Chunk struct {
		Property1 int32 `bitfield:"10"`
		Property2 int32 `bitfield:"19"`
		Property3 int32 `bitfield:"3"`
	}

	tests := []struct {
		value int32
	}{
		{1},
		{-1},
	}

	for i, test := range tests {
		chunk := &Chunk{Property2: value}
		data, err := structex.EncodeByteBuffer(chunk)
		if err != nil {
			t.Fatalf("%d encode: %s", i, err)
		}

		fmt.Printf("%d in value: %d 0x%x encoded 0x%x\n", i, test.value, test.value, data)

		err = structex.DecodeByteBuffer(bytes.NewBuffer(data), chunk)
		if err != nil {
			t.Fatalf("initial decode: %s", err)
		}

		fmt.Printf("%d out value: %d 0x%x\n", i, chunk.Property2, chunk.Property2)
		if test.value != chunk.Property2 {
			t.Fatalf("%d failed: got %d, want %d", i, chunk.Property2, test.value)
		}
	}
}
```

results in:
```
0 in value: 1 0x1 encoded 0x00040000
0 out value: 1 0x1
1 in value: -1 0x-1 encoded 0x00040000
1 out value: 1 0x1
--- FAIL: TestBitfieldSimple (0.00s)
  simple_test.go:48: 1 failed: got 1, want -1
```

--- a/decoder.go
+++ b/decoder.go
@@ -140,7 +140,11 @@ func (d *decoder) readValue(value reflect.Value, tags *tags) (uint64, error) {
 	case reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uint:
 		value.SetUint(v)
 	case reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Int:
-		value.SetInt(int64(v))
+		if v>>(nbits-1) == 1 {
+			value.SetInt(int64(v) - int64((1<<nbits)-1) - 1)
+		} else {
+			value.SetInt(int64(v))
+		}
 	default:
 		return 0, fmt.Errorf("Unsupported read type %s", value.Kind().String())
 	}
@@ -245,6 +249,7 @@ Deserialization occurs according to the annotations in the structure which
 take several options:
 
 Bitfields:
+
 	Bitfields define a structure field with an explicit size in bits. They are
 	analogous to bit fields in the C specification. Un
 
@@ -256,6 +261,7 @@ Bitfields:
 	           bits and should be encoded as zeros.
 
 Dynamic Layouts:
+
 	Many industry standards support dynamically sized return fields where the
 	data layout is self described by other fields. To support such formats
 	two annotations are provided.
@@ -283,6 +289,7 @@ Dynamic Layouts:
 				name Field.
 
 Alignment:
+
 	Annotations can specified the byte-alignment requirement for structure
 	fields. Analogous to the alignas specifier in C. Can only be applied
 	to non-bitfield structure fields.
@@ -291,7 +298,6 @@ Alignment:
 
 	value		An integer value specifying the byte alignment of the field.
 				Invalid non-zero alignments panic.
-
 */
 func Decode(reader io.ByteReader, s interface{}) error {
 

--- a/decoder_test.go
+++ b/decoder_test.go
@@ -25,6 +25,7 @@ package structex
 import (
 	"bytes"
 	"fmt"
+	"math"
 	"math/bits"
 	"testing"
 )
@@ -149,11 +150,11 @@ func TestEndianDecoder(t *testing.T) {
 func TestBitfieldDecoder(t *testing.T) {
 
 	type ts struct {
-		A int `bitfield:"3"`
-		B int `bitfield:"4"`
-		C int `bitfield:"1"`
-		D int `bitfield:"12"`
-		E int `bitfield:"4"`
+		A uint `bitfield:"3"`
+		B uint `bitfield:"4"`
+		C uint `bitfield:"1"`
+		D uint `bitfield:"12"`
+		E uint `bitfield:"4"`
 	}
 
 	var s = new(ts)
@@ -184,9 +185,9 @@ func TestBitfieldDecoder(t *testing.T) {
 func TestNestedDecoder(t *testing.T) {
 
 	type ns struct {
-		M int `bitfield:"3"`
-		N int `bitfield:"4"`
-		O int `bitfield:"1"`
+		M uint `bitfield:"3"`
+		N uint `bitfield:"4"`
+		O uint `bitfield:"1"`
 	}
 
 	type ts struct {
@@ -480,6 +481,39 @@ func TestBoolDecoder(t *testing.T) {
 		}
 		if !s.IsH {
 			t.Errorf("IsH Value Incorrect: Expected: %v Actual: %v", true, s.IsH)
+		}
+	})
+}
+
+func TestSignedDecoder(t *testing.T) {
+	type ts struct {
+		A     int8
+		B     int8  `bitfield:"8"`
+		C     int8  `bitfield:"3"`
+		Dummy uint8 `bitfield:"5"`
+	}
+
+	var s = new(ts)
+
+	var tr = newReader([]byte{-math.MinInt8, -math.MinInt8, 0b11111100})
+
+	unpackAndTest(t, s, tr, func(t *testing.T, i interface{}) {
+		var s = i.(*ts)
+
+		if s.A != math.MinInt8 {
+			t.Errorf("A Value Incorrect: Expected: %d Actual: %d", math.MinInt8, s.A)
+		}
+
+		if s.B != math.MinInt8 {
+			t.Errorf("B Value Incorrect: Expected: %d Actual: %d", math.MinInt8, s.B)
+		}
+
+		if s.C != -4 {
+			t.Errorf("C Value Incorrect: Expected: %d Actual: %d", -4, s.C)
+		}
+
+		if s.Dummy != 0b11111 {
+			t.Errorf("Dummy Value incorrect: Expected: %#b Actual: %#b", 0b111111, s.Dummy)
 		}
 	})
 }

--- a/encoder.go
+++ b/encoder.go
@@ -63,7 +63,7 @@ func (e *encoder) write(value uint64, nbits uint64) error {
 	}
 
 	for nbits != 0 {
-		e.currentByte = uint8(value)
+		e.currentByte = uint8(value & ((1 << nbits)-1))
 		if nbits < 8 {
 			e.bitOffset += nbits
 			return nil

--- a/encoder_test.go
+++ b/encoder_test.go
@@ -26,6 +26,7 @@ package structex
 import (
 	"encoding/binary"
 	"fmt"
+	"math"
 	"testing"
 )
 
@@ -289,13 +290,13 @@ func TestArrayEncoder(t *testing.T) {
 
 func TestArrayLittleEndian16Encoder(t *testing.T) {
 	s := struct {
-		Count uint8 `countOf:"Ts"`
-		Size  uint8 `sizeOf:"Ts"`
+		Count uint8     `countOf:"Ts"`
+		Size  uint8     `sizeOf:"Ts"`
 		Ts    [2]uint16 `little:""`
 	}{
 		Count: 0x00,
 		Size:  0x00,
-		Ts: [2]uint16{0x0102, 0x0304},
+		Ts:    [2]uint16{0x0102, 0x0304},
 	}
 
 	packAndTest(t, s, func(t *testing.T, tw *testWriter) {
@@ -319,13 +320,13 @@ func TestArrayLittleEndian16Encoder(t *testing.T) {
 
 func TestArrayBigEndian16Encoder(t *testing.T) {
 	s := struct {
-		Count uint8 `countOf:"Ts"`
-		Size  uint8 `sizeOf:"Ts"`
+		Count uint8     `countOf:"Ts"`
+		Size  uint8     `sizeOf:"Ts"`
 		Ts    [2]uint16 `big:""`
 	}{
 		Count: 0x00,
 		Size:  0x00,
-		Ts: [2]uint16{0x0102, 0x0304},
+		Ts:    [2]uint16{0x0102, 0x0304},
 	}
 
 	packAndTest(t, s, func(t *testing.T, tw *testWriter) {
@@ -444,6 +445,29 @@ func TestBoolEncoder(t *testing.T) {
 		}
 		if tw.getByte(1) != 0x12 {
 			t.Errorf("Invalid bitfield: Expected: %d Actual: %d", 0x12, tw.getByte(1))
+		}
+	})
+}
+
+func TestSignedEncoder(t *testing.T) {
+	s := struct {
+		A     int8
+		B     int8 `bitfield:"8"`
+		C     int8 `bitfield:"3"`
+		Dummy int8 `bitfield:"5"`
+	}{
+		-1, -1, -1, 0,
+	}
+
+	packAndTest(t, s, func(t *testing.T, tw *testWriter) {
+		if tw.getByte(0) != math.MaxUint8 {
+			t.Errorf("Invalid byte A: Expected: %d Actual: %d", math.MaxUint8, tw.getByte(0))
+		}
+		if tw.getByte(1) != math.MaxUint8 {
+			t.Errorf("Invalid byte B: Expected: %d Actual: %d", math.MaxUint8, tw.getByte(1))
+		}
+		if tw.getByte(2) != 0b00000111 {
+			t.Errorf("Invalid bitfield C: Expected: %#08b Actual: %#08b", 0b00000111, tw.getByte(2))
 		}
 	})
 }

